include/watchdog.h

```diff
@@ -55,40 +71,265 @@ struct wdt_config {
+
+/**
+ * @typedef wdt_api_set_config
+ * @brief Callback API for setting current watchdog configuration.
+ * See wdt_set_config() for argument descriptions. Please note that this
+ * function is deprecated.
+ */
+typedef int (*wdt_api_set_config)(struct device *dev,
+				  struct wdt_config *config);
+
+/**
+ * @typedef wdt_api_reload
+ * @brief Callback API for reloading watchdog.
+ * See wdt_reload() for argument descriptions. Please note that this function
+ * is deprecated.
+ */
 typedef void (*wdt_api_reload)(struct device *dev);
 
+/** @cond INTERNAL_HIDDEN */
 struct wdt_driver_api {
-	wdt_api_enable enable;
+	wdt_api_setup setup;
 	wdt_api_disable disable;
+	wdt_api_install_timeout install_timeout;
+	wdt_api_feed feed;
+	/** Deprecated functions */
+	wdt_api_enable enable;
 	wdt_api_get_config get_config;
 	wdt_api_set_config set_config;
 	wdt_api_reload reload;
 };
+/**
+ * @endcond
+ */
 
-static inline void wdt_enable(struct device *dev)
+/**
+ * @brief Set up watchdog instance.
+ *
+ * This function is used for configuring global watchdog settings that
+ * affect all timeouts. It should be called after installing timeouts.
+ * After successful return, all installed timeouts are valid and must be
+ * serviced periodically by callig wdt_feed().
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param options Configuration options as defined by the WDT_OPT_* constants
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP If any of the set options is not supported.
+ * @retval -EBUSY If watchdog instance has been already setup.
+ */
+static inline int wdt_setup(struct device *dev, u8_t options)
 {
 	const struct wdt_driver_api *api =
 		(const struct wdt_driver_api *)dev->driver_api;
 
-	api->enable(dev);
+	return api->setup(dev, options);
 }
 
-static inline void wdt_disable(struct device *dev)
+/**
+ * @brief Disable watchdog instance.
+ *
+ * This function disables the watchdog instance and automatically uninstalls all
+ * timeouts. To set up a new watchdog, install timeouts and call wdt_setup()
+ * again. Not all watchdogs can be restarted after they are disabled.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval -EFAULT If watchdog instance is not enabled.
+ * @retval -EPERM If watchdog can not be disabled directly by application code.
+ */
+static inline int wdt_disable(struct device *dev)
 {
 	const struct wdt_driver_api *api =
 		(const struct wdt_driver_api *)dev->driver_api;
 
-	api->disable(dev);
+	return api->disable(dev);
 }
 
-static inline void wdt_get_config(struct device *dev,
+/**
+ * @brief Install new timeout.
+ *
+ * This function must be used before wdt_setup(). Changes applied here
+ * have no effects until wdt_setup() is called.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cfg Pointer to timeout configuration structure.
+ *
+ * @retval 0 If successful.
```

Comment:
It looks like we forgot to update the documentation to inform the user that the function returns channel_id. This makes usage of this API confusing. It should probably be fixed for 1.12 release.